[Email Editor] Generate email editor post for WooCommerce transactional emails

triple0t opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

Closes #56109 

### Email Editor Block Templates Implementation

This PR introduces block-based email templates for WooCommerce transactional emails, improving the customization experience and making email templates more maintainable.

#### Changes

- Added initial block-based email templates for all core WooCommerce transactional emails
- Implemented personalization tags for dynamic content replacement
- Refactored email template generation to use transients for version tracking
- Enhanced email content handling and editor integration
- Improved structure and dynamic content handling in email templates
- Added support for block-based email customization
- Removed the top-level `WP Admin > Woo Emails` menu
- Removed the template selector from the Email editor settings sidebar panel
- Update Woo Content block placeholder preview

#### Email Templates Added

- Admin emails:
  - New order
  - Failed order
- Customer emails:
  - New account
  - Reset password
  - Order confirmation
  - Processing order
  - Completed order
  - Failed order
  - On-hold order
  - Refunded order
  - Customer note
  - Invoice

#### Technical Details

- All new templates use the block editor format with proper WordPress block syntax
- For this PR, we generate the initial email posts when `/wp-admin/admin.php?page=wc-settings&tab=email` page is visited
- Initial block email content templates are stored in `plugins/woocommerce/templates/emails/block/`
- We use `plugins/woocommerce/templates/emails/block/general-block-email.php` to render the dynamic part of the email template
- Added support for personalization tags using the `<!--[woocommerce/tag-name]-->` syntax
- Templates maintain backward compatibility with existing email template structure
- Email posts are associated with WooCommerce transactional emails with the option: `woocommerce_email_templates_WC-EMAIL-ID_post_id`
- We ensure the templates are generated with transient-based version tracking. option: `wc_email_editor_initial_templates_generated`
- Version 9.9.0 targeted for these changes

### Screenshots or screen recordings:

<!-- If this PR includes UI changes, please provide screenshots or a screen recording for clarity. -->
<!-- This section can be removed if not applicable. -->

**Email editor**
![Screenshot 2025-03-20 at 7 45 29 PM](https://github.com/user-attachments/assets/48b9da0b-96fd-4035-ba9c-caeaf2d237fa)
![Screenshot 2025-03-20 at 7 45 39 PM](https://github.com/user-attachments/assets/d953b1f7-5301-4c89-99d1-13835fcb121d)

**Email client**
![Screenshot 2025-03-20 at 7 49 41 PM](https://github.com/user-attachments/assets/818ac63b-4a8f-43e9-8354-45f6615e4c3e)

![Screenshot 2025-03-20 at 9 30 14 PM](https://github.com/user-attachments/assets/e0ba8fd1-afa3-4b2c-90b5-07a6587b7629)




<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

NOTE: Testing this PR may be a little difficult without the `Woo emails` UI. You can temporarily add it back for your test.
Add
```php
  'public'       => false,
  'show_ui'      => true, 
  'show_in_menu' => true,
  'show_in_rest' => true,
```
to https://github.com/woocommerce/woocommerce/blob/b647643298086579dc6a60c9e1a60065c945a5aa/plugins/woocommerce/src/Internal/EmailEditor/Integration.php#L98-L109

1. Enable the `block_email_editor` experimental flag (in DB, set `woocommerce_feature_block_email_editor_enabled` option to `yes` Or using wp-cli `wp option add woocommerce_feature_block_email_editor_enabled yes`).
2. Visit `/wp-admin/admin.php?page=wc-settings&tab=email` for automatic email post-generation
3. Create a new order to test admin new order email
4. Process an order through different statuses to test customer emails
5. Add customer notes to verify note emails
6. Test account-related emails (new account, reset password)
7. Verify personalization tags are properly replaced in all emails
8. Check email template customization in the block editor


To retest template generation
* Delete transient `wp transient delete wc_email_editor_initial_templates_generated`
* Delete all or some email posts association with `SELECT * FROM `wp_options` WHERE `option_name` LIKE '%woocommerce_email_templates_%'` 
* Delete the required email editor posts
* Visit `/wp-admin/admin.php?page=wc-settings&tab=email` for automatic email post-generation
<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

triple0t committed:
Add WCTransactionalEmails functionality to Email Editor

- Introduced WCTransactionalEmails class to manage transactional email templates.
- Added WCEmailTemplateGenerator for generating default email templates.
- Implemented WCEmailTemplateManager for handling email template storage and retrieval.
- Updated EmailEditorServiceProvider and Integration classes to include new transactional email components.
- Modified BlockEmailRenderer to support multiple post statuses for email templates.

triple0t committed:
Add initial block email templates for various order statuses

- Introduced new email templates for admin and customer notifications, including cancelled, failed, new, completed, on-hold, processing, refunded, and reset password emails.
- Each template includes structured content and placeholders for dynamic data, ensuring a consistent user experience.
- Enhanced the email editor's capabilities by providing block-based templates for better customization and management.

triple0t committed:
Add block editor email template content to various email classes

- Implemented a new method `get_block_editor_email_template_content` in multiple email classes, including cancelled order, customer new account, customer note, customer refunded order, customer reset password, failed order, and new order.
- This method retrieves the block editor email template content, enhancing the customization and management of email notifications.
- Updated the email content processing to utilize the new block editor templates for improved user experience.

triple0t committed:
Refactor email templates to use placeholders for dynamic content

- Updated various email templates to replace direct order and customer data retrieval with placeholders for dynamic content.
- Enhanced consistency across email templates by utilizing the block editor's placeholder system for order numbers, customer names, and store details.
- Improved readability and maintainability of email templates by streamlining the PHP code structure.

triple0t committed:
Refactor email template generation to utilize core transactional emails

- Updated the WCEmailTemplateGenerator to dynamically initialize default email templates based on core transactional emails.
- Implemented a method to retrieve email templates, enhancing maintainability and reducing hardcoded values.
- Improved the generation of initial email templates by checking for existing templates and only creating those that are missing.
- Added detailed documentation for new methods to clarify their purpose and usage.

triple0t committed:
Add personalization tags for shopper and order details in email templates

- Introduced new personalization tags for shopper's first name, last name, full name, and username, enhancing the dynamic content capabilities of email templates.
- Added order-related personalization tags, including order number and admin order note, to improve the customization of transactional emails.
- Updated BlockEmailRenderer to include the wc_email context for better data handling in email generation.

triple0t committed:
Enhance BlockEmailRenderer and WCEmailTemplateManager for improved email template handling

- Added WCEmailTemplateManager to manage email template retrieval and existence checks, streamlining the process of accessing email posts.
- Updated BlockEmailRenderer to utilize the new template manager for fetching email posts, improving code maintainability and readability.
- Introduced detailed documentation for new methods in WCEmailTemplateManager to clarify their functionality.

triple0t committed:
Refactor email templates for improved structure and dynamic content handling

- Updated WCEmailTemplateGenerator to filter and include only core transactional emails, enhancing template management.
- Modified various email templates to improve alignment and structure, including centering text and adjusting HTML elements for better readability.
- Enhanced customer account-related emails to utilize data attributes for links, improving accessibility and functionality.
- Implemented locking features in customer note templates to prevent unwanted modifications, ensuring template integrity.

triple0t committed:
Fix Personalization Tag not loading for order note

triple0t committed:
Fix PHPCS violations in updated files

triple0t committed:
Refactor email preview handling and enhance content processing

- Introduced WooContentProcessor for improved email content rendering and CSS preparation.
- Updated EmailPreview class to expose the email object and public methods for setting up and cleaning up filters.
- Modified BlockEmailRenderer to utilize the new WooContentProcessor for CSS handling, enhancing email rendering consistency.
- Removed deprecated capture_woo_content method to streamline the email preview process.

triple0t committed:
Enhance email content handling and improve editor integration

- Added a minimum height to the email content placeholder for better visibility in the editor.
- Enhanced the EmailPreview class with additional user properties for improved email setup.
- Implemented a check in WooContentProcessor to return an empty string for empty content, preventing unnecessary processing and css_inliner error.

triple0t committed:
Refactor email editor integration and remove sidebar extension

- Updated the send button label in the email editor to use the translation function for better localization support.
- Removed the sidebar extension component and its associated functionality to streamline the email editor integration.

triple0t committed:
Refactor email editor integration and improve content handling

- Replaced useEntityRecord with useSelect for better state management in the email content placeholder.
- Introduced strict types declaration in the WC_Admin class for better type safety.
- Enhanced error handling in the email preview generation process to prevent crashes and improve user experience.

triple0t committed:
Remove temporary UI visibility settings from email post type registration in the email editor integration

triple0t committed:
Refactor initial email template generation to use transients for version tracking

- Updated the method to check for initial email templates using a transient instead of an option, improving performance and reducing database load.
- Set the transient to the current WooCommerce version, ensuring templates are only generated once per version.

triple0t committed:
Add change log file

triple0t committed:
Add functionality to delete email templates association upon email editor post deletion

- Introduced a new method in the Integration class to delete the associated email template when an email editor post is permanently deleted.
- Updated the WCEmailTemplateManager to include a method for deleting the post ID of a specific email template type, ensuring proper cleanup of email templates.

triple0t committed:
Update email templates to use correct shortcode for account link

- Changed the placeholder for the "My account" link in the new account email templates to use the correct shortcode format, ensuring proper rendering in emails.

costasovo commented on plugins/woocommerce/src/Internal/EmailEditor/WCTransactionalEmails/WCEmailTemplateGenerator.php:
```diff
@@ -0,0 +1,150 @@
+		$this->template_manager = WCEmailTemplateManager::get_instance();
+	}
+
+	public function init() {
```
I would suggest using a different method name. AFAIK the init will be triggered in case we grab the service from DI, so I find it potentially dangerous.

triple0t replied:
Updated [here](https://github.com/woocommerce/woocommerce/pull/56567/commits/75a6a46927eedc62c2fd6aac6c02ba4e538b181b)

costasovo commented on plugins/woocommerce/src/Internal/EmailEditor/WCTransactionalEmails/WCEmailTemplateManager.php:
```diff
@@ -0,0 +1,46 @@
+
+use Automattic\WooCommerce\Internal\EmailEditor\Integration;
+
+class WCEmailTemplateManager {
```
I would suggest a different name. It is a sort of template for the content, but I think this might be easily confused with the main template, which carries the header and footer. What about `WCTransactionalEmailPostsManager` or something similar?

triple0t replied:
Done in https://github.com/woocommerce/woocommerce/pull/56567/commits/3ac5319bdce5c108d4d1d682e0bc9569e173eb8e

costasovo commented on plugins/woocommerce/src/Internal/EmailEditor/WCTransactionalEmails/WCEmailTemplateGenerator.php:
```diff
@@ -0,0 +1,150 @@
+use Automattic\WooCommerce\Internal\EmailEditor\BlockEmailRenderer;
+use Automattic\WooCommerce\Internal\EmailEditor\EmailTemplates\WooEmailTemplate;
+
+class WCEmailTemplateGenerator {
```
What about `WCTransactionalEmailPostsGenerator` to avoid confusion with the wp_template posts?

triple0t replied:
Done in https://github.com/woocommerce/woocommerce/pull/56567/commits/3ac5319bdce5c108d4d1d682e0bc9569e173eb8e

costasovo commented on plugins/woocommerce/includes/emails/class-wc-email-cancelled-order.php:
```diff
@@ -142,6 +142,23 @@ public function get_content_plain() {
+		 *
+		 * @return string
+		 */
+		public function get_block_editor_email_template_content() {
```
I'm not sure this is a good idea. This will force changes in emails added by third parties who respect the WC_Email class ([example](https://github.com/woocommerce/woocommerce-subscriptions/blob/c7da0607a98681f555925126d2445e80b18e32aa/includes/payment-retry/emails/class-wcs-email-payment-retry.php#L17)) to add this method for the support.

What do we want to achieve here? The commit message doesn't give much insight here. Can't we do it in the patent WC_Email class?

triple0t replied:
> I'm not sure this is a good idea. This will force changes in emails added by third parties who respect the WC_Email class ([example](https://github.com/woocommerce/woocommerce-subscriptions/blob/c7da0607a98681f555925126d2445e80b18e32aa/includes/payment-retry/emails/class-wcs-email-payment-retry.php#L17)) to add this method for the support.

Not really. It will not affect third-party emails without the email editor posts.

> What do we want to achieve here? The commit message doesn't give much insight here. Can't we do it in the patent WC_Email class?

We execute this [method](https://github.com/woocommerce/woocommerce/blob/34bc33ca0ed53913541cb1a6e5c69b57c87c424b/plugins/woocommerce/src/Internal/EmailEditor/WooContentProcessor.php#L110) when we want to get only the [order details part](https://github.com/woocommerce/woocommerce/blob/34bc33ca0ed53913541cb1a6e5c69b57c87c424b/plugins/woocommerce/templates/emails/block/general-block-email.php) of the template for the Woo content placeholder.

costasovo replied:
@triple0t Thanks for explaining! I think I understand now.

I thought we could achieve something similar by modifying the original email templates where we could use a feature switch and comment and skip the text parts we want to move to blocks and keep only the main content. But I see we would get basically what is now the general email content in many cases.

My only concern now is what happens if a third party extends the WC_Email class and is listed in WC_Emails. We will create the default content post from the fallback block content, right? So it will have a post. Then when sending we will replace the Woo_block with the general email content from the main WC_Email:get_block_editor_email_template_content, and we will ignore the third-party HTML template. 
So basically, a third party needs to provide a block template to be able to add some customized content. Or am I missing something? 

I thought that with the approach mentioned above, we could still grab the full content from the HTML templates (including text parts, etc.). from third-party emails and use that to replace the Woo placeholder block. That way, we could keep sending reasonable content also for the email added by a third party.

triple0t replied:
> I thought we could achieve something similar by modifying the original email templates where we could use a feature switch and comment and skip the text parts we want to move to blocks and keep only the main content. But I see we would get basically what is now the general email content in many cases.

I thought of that approach as well but didn't like it because it's not very clean.
It has the following limitations
1. We have no control over the email content when the user has overridden the transaction email template
2. When a user has overridden the transactional email template, what we set and display on the email editor may differ from what is rendered in the email client.
3. It makes extending the WooCommerce email templates complicated. For example, now that you have multiple if-else statements, you don't know what they are used for or if you even need them in your custom template. 
4. It makes it difficult for us to implement the project's second phase, which is creating a block version of the WooCommerce order details. We would still have to do something similar to what I've done here.

> My only concern now is what happens if a third party extends the WC_Email class and is listed in WC_Emails. We will create the default content post from the fallback block content, right?

No, we don't create the email post by default. Unless the extender adds the email ID to the required list using the `woocommerce_transactional_emails_for_block_editor` filter. Steps [here](https://github.com/woocommerce/woocommerce/pull/56567#issuecomment-2747115685).
That is only when we create the email post and use the fallback block content when not provided.

costasovo replied:
> I thought of that approach as well but didn't like it because it's not very clean.
It has the following limitations

@triple0t Thanks for the explanation and the reasons. I forgot about the template override, and it makes perfect sense for the second phase and overall, your approach seems more reasonable 👍 

> No, we don't create the email post by default. Unless the extender adds the email ID to the required list using the woocommerce_transactional_emails_for_block_editor filter. Steps https://github.com/woocommerce/woocommerce/pull/56567#issuecomment-2747115685.
That is only when we create the email post and use the fallback block content when not provided.

Oh! Sorry, I missed that. So, it will be opt-in for the third party. Nice! Now it clicked! So, in that case I think that in the new email listing, the emails without the post should still open the old email manage screen. I can update it in my branch. What do you think?

Sorry for the many questions, and thanks for your patience! I wanted to properly understand the overall idea. I could not make or failed to make the full picture properly from the code and commit messages.

triple0t replied:
> Oh! Sorry, I missed that. So, it will be opt-in for the third party.

@costasovo, Yes, that is correct.

> So, in that case I think that in the new email listing, the emails without the post should still open the old email manage screen. I can update it in my branch. What do you think?

I think that's fine. It would make it easy to still manage other emails.

costasovo commented on plugins/woocommerce/src/Internal/EmailEditor/WCTransactionalEmails/WCEmailTemplateGenerator.php:
```diff
@@ -14,60 +14,145 @@ class WCEmailTemplateGenerator {
+		$email_types = $wc_emails->get_emails();
+		$email_types = array_filter( $email_types, function( $email ) use ( $core_transactional_emails ) {
+			return in_array( $email->id, $core_transactional_emails );
+		} );
```
Just an idea. I think we could also support 3rd party emails by having universal email content that would contain only the woo_placeholder block. So when generating the email posts, we could check if there is a post content for the type, and if not, we would use the universal one. We could then allow 3rd party integrators to provide their own block content via a filter.
For now, I think core emails are fine, but this way, we can add basic support for correctly registered 3rd party emails soon.

triple0t replied:
I share the same thoughts here. I would love to add a filter here to support creating initial email posts for third-party extensions, but I recall Pavel asking us to hold off on doing that for now. I may add it, but mark it as experimental.

triple0t replied:
I added support for loading an initial block email content [here](https://github.com/woocommerce/woocommerce/pull/56567/commits/8f42f47f268afdefb2a36382e05e4364dd98250b) when not provided by the theme/plugin developer and added a [filter](https://github.com/woocommerce/woocommerce/pull/56567/commits/d9cf3bd20d4013f88a4b52766364afe876311c43#) for including an email ID when fetching the list of transactional emails.

costasovo commented on plugins/woocommerce/src/Internal/EmailEditor/WCTransactionalEmails/WCEmailTemplateGenerator.php:
```diff
@@ -80,15 +165,19 @@ public function generate_email_templates() {
 		$post_data = [
 			'post_type' => Integration::EMAIL_POST_TYPE,
-			'post_status' => 'publish',
+			'post_status' => $template_data['enabled'] ? 'publish' : 'draft',
```
I see! That's why the `Order details` aka `customer_invoice ` email is a draft. This one is manually sent so it will never be set as enabled. You can test if the email is manual by calling `WC_Email::is_manual` then perhaps the logic could be `$template_data['enabled'] || $template_data['manual'] ? 'publish' : 'draft'`

triple0t replied:
Noted. Will do.

triple0t replied:
Done in https://github.com/woocommerce/woocommerce/pull/56567/commits/d9cf3bd20d4013f88a4b52766364afe876311c43

triple0t commented:
Thank you for the quick review, @costasovo 

>  I noticed that the "Order details" email is draft all the other emails have state publish. Is that on purpose? The "Order details" email is triggered manually from the order's admin screen. It can't be enabled or disabled. I would expect it being published as well.

That's true. We aren't doing anything with the email post status right now, so we can enable all emails. 
The idea is that if a user intentionally disables a WooCommerce transactional email, the email post should reflect that status. I think we are planning to do that in another issue so I will update this PR.

>  I've experienced some issues with the generating because initially, I had some leftovers in the options table, and I had woocommerce_email_templates_customer_processing_order_post_id, which pointed to a nonexistent email. The email was skipped and was not generated. Perhaps we could make the generating more robust and check whether the records point to existing woo_email posts. What do you think?

I initially implemented this to check if the email post exists but switched it to using the options table for performance reasons. It wasn't very practical to run get_post 10 to 12 times. I added a [hook](https://github.com/woocommerce/woocommerce/blob/34bc33ca0ed53913541cb1a6e5c69b57c87c424b/plugins/woocommerce/src/Internal/EmailEditor/Integration.php#L86) to notify us when a post is deleted and to perform options value [cleanup](https://github.com/woocommerce/woocommerce/blob/34bc33ca0ed53913541cb1a6e5c69b57c87c424b/plugins/woocommerce/src/Internal/EmailEditor/Integration.php#L164) on post delete.
We shouldn't experience the same issue again.


> I assume that's expected behavior, as it might not make sense to send an email for some transitions.

I believe so, yes.

>  I was thinking that in the future, we may put the generating process into a background job, so that it is not dependent on visiting the emails page. What do you think?

Yes, please. I agree with you.

costasovo commented on plugins/woocommerce/includes/admin/class-wc-admin.php:
```diff
@@ -7,6 +7,8 @@
  * @version  2.6.0
  */
 
+declare(strict_types=1);
```
Is that intentional? I have a concern that in such an old class we may run into issues because some variables in WP tend to be sometimes string sometimse an integer...

triple0t replied:
Yes... but we don't have any type-hint here, so it should be fine.

costasovo reviewed (commented):
@triple0t Wow, such an incredible amount of work and functionality added! Great job! I think it is good overall. I just have a couple of questions and suggestions.

I also think it would be great to add some unit tests for the classes that generate the emails.

triple0t committed:
Rename WCEmailTemplateGenerator and WCEmailTemplateManager classes

- Replaced the WCEmailTemplateManager with WCTransactionalEmailPostsManager in the BlockEmailRenderer and Integration classes for improved email template handling.
- Added WCTransactionalEmailPostsGenerator to manage the generation and initialization of WooCommerce transactional email templates.
- Updated the WCTransactionalEmails class to utilize the new generator for initializing email templates, enhancing the overall email management system.

triple0t committed:
Enhance WCTransactionalEmailPostsGenerator and WCTransactionalEmails classes

- Added transient tracking for initial email template generation to prevent redundant processing.
- Updated the method for fetching core transactional emails to utilize a filter, improving flexibility.
- Refactored email template generation logic to check for enabled status, ensuring proper post status assignment.
- Improved initialization of email templates to run only on relevant WooCommerce admin pages, enhancing performance and user experience.

triple0t committed:
Refactor WCTransactionalEmailPostsGenerator for improved email template handling

- Updated type hints and variable annotations for better code clarity and type safety.
- Simplified the email template generation logic by directly using email data instead of an array structure.
- Enhanced the method for generating email templates to ensure proper handling of enabled and manual statuses.
- Improved variable naming for better readability and understanding of the code flow.

triple0t committed:
Ensure the Email editor loads a default initial block content when it is not provided by the plugin/theme developer

- Added error handling for email template generation to manage exceptions and ensure graceful degradation.
- Introduced a default block content template to be used when a custom template is not available, enhancing user experience and reliability.
- Utilized StringUtil for improved error detection in template rendering, ensuring robust handling of missing files.

triple0t committed:
Refactor WCTransactionalEmailPostsGenerator and related tests for improved clarity and functionality

- Renamed the `init` method to `initialize` for better semantic clarity.
- Updated the `init_default_templates` method to public for broader accessibility.
- Adjusted return values in the email generation logic to enhance error handling.
- Added comprehensive unit tests for `WCTransactionalEmailPostsGenerator`, `WCTransactionalEmailPostsManager`, and `WCTransactionalEmails` to ensure robust functionality and reliability.

triple0t commented:
Thank you for the in-depth review, @costasovo 
I have addressed your feedback and added more test cases to the PR.

**Updates**
* Renamed the WCEmailTemplateGenerator and WCEmailTemplateManager classes
* Added a filter for including an email ID when fetching the list of transactional emails. (`woocommerce_transactional_emails_for_block_editor`)
* Allow the process of generating the initial block email contents to run from any WooCommerce admin pages
* Refactor WCTransactionalEmailPostsGenerator class to only generate the email content when required compared to the previous implementation where all emails were generated on init.
* Added support for loading an initial block email content when not provided in `theme/woocommerce/emails/block/email-id.php`
* Add more test cases.

**Third-party Integration/Extension** 

The following steps must be taken to allow the block editor to generate and manage extension emails.
1. Ensure the email is registered using the `woocommerce_email_classes` filter
2. Add the email ID to the required list using the `woocommerce_transactional_emails_for_block_editor` filter
3. Add an initial block email content to `theme/woocommerce/emails/block/email-id.php` (content should be a valid WordPress Block syntax)
4. Clear the transient `wc_email_editor_initial_templates_generated` and visit any WooCommerce admin page.

Note: Block content customization is done using Personalization tags; examples [here](https://github.com/woocommerce/woocommerce/tree/75a6a46927eedc62c2fd6aac6c02ba4e538b181b/plugins/woocommerce/templates/emails/block). We are still in the process of building the integration points for third-party Personalization tag integration, i.e., allowing extension developers to register custom Personalization tags for their own email contents.

costasovo commented on plugins/woocommerce/src/Internal/EmailEditor/WCTransactionalEmails/WCTransactionalEmailPostsGenerator.php:
```diff
@@ -109,15 +110,41 @@ function ( $acc, $email ) {
+					'plain_text'    => false,
+					'email'         => $email,
+				)
+			);
```
Do we need to provide these data? At the time of generating email posts we use personalized tags. What order are we passing there?

triple0t replied:
You're right. We don't need this here. Removed in https://github.com/woocommerce/woocommerce/pull/56567/commits/759a1c1fc95f0748344f7d71eb8e001521285545

costasovo commented on plugins/woocommerce/src/Internal/EmailEditor/WCTransactionalEmails/WCTransactionalEmailPostsGenerator.php:
```diff
@@ -72,7 +72,7 @@ final public function init() {
 	 * It fetches all the emails from WooCommerce and filters them to include only the core transactional emails.
 	 */
-	private function init_default_templates() {
+	public function init_default_templates() {
```
Could we use `init_default_transactional_emails` or `init_default_email_contents` or just something else but templates? I see it is used in many other places. Theoretically, it is a template for the content. Maybe we could at least say they are content templates. I would like to not use templates with this context to avoid confusion with the main email template. In the future, we will add multiple email templates and I'm afraid it will be confusing what a template is.

triple0t replied:
Noted. Implemented here https://github.com/woocommerce/woocommerce/pull/56567/commits/759a1c1fc95f0748344f7d71eb8e001521285545

costasovo reviewed (commented):
@triple0t Thank you for the updates! Kudos for adding the fallback post content!

I have a couple of additional questions. Please take a look at the comments.

triple0t committed:
Remove blogname parameter from the block email data in various email classes because it is unused

triple0t committed:
Use constant for _wc_email_type and switch `WOOCOMMERCE_VERSION` to `Constants::get_constant('WC_VERSION')`

This improves the readability of the code and allows more extension

triple0t committed:
Update WCTransactionalEmailPostsGenerator to use methods for title and description retrieval

- Changed `post_title` to use `get_title()` method for better encapsulation.
- Added `post_excerpt` to utilize `get_description()` method, enhancing email data handling.

costasovo commented on plugins/woocommerce/src/Internal/EmailEditor/WooContentProcessor.php:
```diff
@@ -88,7 +107,7 @@ private function capture_woo_content( \WC_Email $wc_email ): string {
 		remove_all_filters( 'woocommerce_email_footer' );
 
-		$woo_content = $wc_email->get_content_html();
+		$woo_content = $wc_email->get_block_editor_email_template_content();
```
After the previous conversation I realized that now, with the new `get_block_editor_email_template_content` and copied template with the main content we no longer need to disable and re-enable `woocommerce_email_header` and `woocommerce_email_footer` filters because they are no longer included in the template that is generated via `get_block_editor_email_template_content`. We actually don't event need the method as all we need to call is `$wc_email->get_block_editor_email_template_content();`

I tested it locally and it seems to work without them. This is a nice side-effect of your idea with the copy of general email content. 👍 

I'm sorry that I didn't realize it in the first or second review. Could you please remove the manipulation with filters?

triple0t replied:
No problem. I missed this during the last cleanup as well. Done in https://github.com/woocommerce/woocommerce/pull/56567/commits/472982c792130e4a62e92253f593a64595afa7d6

costasovo commented on plugins/woocommerce/src/Internal/EmailEditor/WCTransactionalEmails/WCTransactionalEmailPostsGenerator.php:
```diff
@@ -0,0 +1,273 @@
+			$default_template_name = 'emails/block/default-block-content.php';
+			$template_html         = wc_get_template_html(
+				$default_template_name,
+				array(
+					'order'         => $email->object,
+					'sent_to_admin' => true,
+					'plain_text'    => false,
+					'email'         => $email,
+				)
```
I missed this place before. I assume we can use an empty array here the same as the above.

triple0t replied:
Thanks for catching it. Done in https://github.com/woocommerce/woocommerce/pull/56567/commits/472982c792130e4a62e92253f593a64595afa7d6

costasovo reviewed (commented):
@triple0t, Thank you for addressing my comments and for your patience! 
I have two more comments with tiny issues, but I think we should rather address them now so that we don't forget.

triple0t committed:
Clean up: remove unnecessary information from the methods

costasovo reviewed (approved):
@triple0t Great job!

I performed another round of manual testing and it works as expected.

Here is a screenshot of the generated emails:

<img width="1510" alt="Screenshot 2025-03-25 at 16 52 22" src="https://github.com/user-attachments/assets/0639507f-3c55-4231-8685-1cd28e7efeba" />

Here is the recording of testing various emails:

https://github.com/user-attachments/assets/f0bcafb7-3786-4197-bd0d-c9a83ff4994b


I see there are a couple of jobs failing. I guess a rebase might help with that.
